website/server/libs/groupTasks.js

```diff
@@ -1,61 +1,268 @@
+          if (masterTask.history[i].userId === groupMemberTask.userId) {
+            masterTask.history.splice(i, 1);
+            break;
+          }
+        }
+      }
+    }
+
+    promises.push(masterTask.save());
+    await Promise.all(promises);
+  }
 }
 
-async function _evaluateAllAssignedCompletion (masterTask) {
+async function _evaluateAllAssignedCompletion (masterTask, groupMemberTask) {
   let completions;
   if (masterTask.group.approval && masterTask.group.approval.required) {
     completions = await Tasks.Task.count({
       'group.taskId': masterTask._id,
       'group.approval.approved': true,
     }).exec();
-    completions += 1;
+    // Since an approval is not yet saved into the group member task, count it
+    // But do not recount a group member completing an already approved task
+    if (!groupMemberTask.completed) completions += 1;
   } else {
     completions = await Tasks.Task.count({
       'group.taskId': masterTask._id,
       completed: true,
     }).exec();
   }
-  if (completions >= masterTask.group.assignedUsers.length) {
-    await _completeMasterTask(masterTask);
+  await _completeOrUncompleteMasterTask(
+    masterTask,
+    completions >= masterTask.group.assignedUsers.length,
+  );
+}
+
+async function groupTaskCompleted (groupMemberTask, user, now) {
+  const masterTask = await Tasks.Task.findOne({
+    _id: groupMemberTask.group.taskId,
+  }).exec();
+
+  if (!masterTask || !masterTask.group) return false;
+
+  if (masterTask.history && masterTask.history.length > 0) {
+    // I hope we don't really need to loop the history and check every date
+    // (User with a late day start in Australia vs and early one in Hawaii?)
+    const taskLastHistory = masterTask.history[masterTask.history.length - 1];
+    // Check if the history entry has a user who completed the task
+    // TODO When group scoring is implemented, add history entries for group missed dailies
+    if (taskLastHistory.userId) {
+      const lastCompletingUser = await Users.findById(taskLastHistory.userId, 'preferences');
+      if (lastCompletingUser) {
+        // Check what the completing user's "day" was when the task was completed
+        const taskLastCompletedDay = startOfDay(
+          defaults(
+            { now: moment(taskLastHistory.date) },
+            lastCompletingUser.preferences.toObject(),
+          ),
+        );
+        const userDay = startOfDay(defaults({ now }, user.preferences.toObject()));
+        if (userDay.isSame(taskLastCompletedDay)) return true;
+      }
+    }
   }
+
+  // Did not find a matching completion or valid completing user
+  return false;
 }
 
 async function handleSharedCompletion (groupMemberTask) {
   const masterTask = await Tasks.Task.findOne({
     _id: groupMemberTask.group.taskId,
   }).exec();
 
-  if (!masterTask || !masterTask.group || masterTask.type !== 'todo') return;
+  if (!masterTask || !masterTask.group || masterTask.type === 'habit') return;
```

Comment:
should we also exclude rewards since `handleSharedCompletion ` seems to be called for all tasks inclusing group rewards?